docs: add troubleshooting guide (#6371)
strongloop/loopback-next: docs/site/Troubleshooting.md

---
lang: en
title: 'Troubleshooting'
keywords: LoopBack 4.0, LoopBack 4, Node.js, TypeScript, Debug
sidebar: lb4_sidebar
permalink: /doc/en/lb4/Troubleshooting.html
---

In general, when you are developing an application, use the `npm` command to run
it. This enables you to see stack traces and console output immediately.

For example:

```
$ cd myapp
$ npm start
```

LoopBack 4 also provides multiple ways to help you verify or diagnose the app.

> allows you to specify debug strings that the application will display to the

Author: weird, removed it, it's a dup of the next paragraph.

[Binding documentation page](https://loopback.io/doc/en/lb4/Binding.html).

## Updating Compiled Files

The TypeScript application is compiled to JavaScript files and runs. Therefore
forgetting to rebuild the updated TS files could run into errors.

For example, people deleted a Controller file but running `npm start` still
keeps complaining about it. This error implies the compiled JS files probably
didn't get updated.

To fix it, run `npm run clean` to delete all the compiled files. By doing so, it
forces the build to generate the compiled JS files next time when you start the
application using `npm start` or call `npm run build`.

## Debugging with Visual Studio Code

LoopBack team develops with editor Visual Studio Code in which we defined some
tasks to help people debug their projects.

### Debugging Packages

Project `loopback-next` contains a folder `.vscode` at root level for debugging
the test files. When start the debug task, it will stop at any breakpoint
triggered by the tests.

- Run Mocha tests: Debug all tests inside the `/packages` folder.
- Debug Current Test File: Only debug the tests inside the current open file.

### Debugging Applications and Examples

> Each application created by `lb4 app` contains a default "Launch Program" task

Reviewer: are you referring to the `launch.json` under `.vscode` folder? I just created an app using `lb4 app`, didn't see it created by the tooling.
I think it may not be a bad idea to have the `launch.json` created when scaffolding an app, but it's out of scope of this PR.

Author: hmm, I was checking with the example repos, wasn't aware of the created app doesn't really has it, will remove.

---
lang: en
title: 'Troubleshooting'
keywords: LoopBack 4.0, LoopBack 4, Node.js, TypeScript, Debug
sidebar: lb4_sidebar
permalink: /doc/en/lb4/Troubleshooting.html
---

In general, when you are developing an application, use the `npm` command to run
it. This enables you to see stack traces and console output immediately.

For example:

```
$ cd myapp
$ npm start
```

LoopBack 4 also provides multiple ways to help you verify or diagnose the app.

> 

Author: @dhmlau We already have separate pages for them, if there are any tips, they should be added there I think. The section in "Troubleshooting.md" should stay brief.

> 

Reviewer: What I meant was to have "general tips" and "troubleshooting for specific errors" sections. Currently the levels of abstraction of the sections in this page don't quite match.

## Setting Debug Strings

If you get an error but it doesn't provide enough information, you can turn on
debug mode by starting your application using `DEBUG=<DEBUG_STRING> npm start`.
[Setting debug strings](Setting-debug-strings.md) for details.

## Running Tests with Mocha

You can debug tests by running mocha commands as documented on page
[Debugging with Mocha tests](Debugging-tests-with-mocha.md).

> ## Resolving Common Seen Errors

Author: I will use `commonly found errors`

---
lang: en
title: 'Troubleshooting'
keywords: LoopBack 4.0, LoopBack 4, Node.js, TypeScript, Debug
sidebar: lb4_sidebar
permalink: /doc/en/lb4/Troubleshooting.html
---

In general, when you are developing an application, use the `npm` command to run
it. This enables you to see stack traces and console output immediately.

For example:

```
$ cd myapp
$ npm start
```

LoopBack 4 also provides multiple ways to help you verify or diagnose the app.

## Setting Debug Strings

If you get an error but it doesn't provide enough information, you can turn on
debug mode by starting your application using `DEBUG=<DEBUG_STRING> npm start`.

> [Setting debug strings](Setting-debug-strings.md) for details.

Reviewer 2: ```suggestion
See [Setting debug strings](Setting-debug-strings.md) for details.
```

re-running the application. By default, if you start the application using
`npm start`, it rebuilds the project for you.

In some cases, the compiled JS files might not get updated properly even you
build the project. For example, you deleted a Controller file but still get
errors about the removed Controller during application startup, it implies the
compiled JS files probably didn't get updated.

To fix it, run `npm run clean` to delete all the compiled files. By doing so, it
forces the build to generate the compiled JS files next time when you start the
application using `npm start` or call `npm run build`.

## Submitting Sample Application for Problem Determination

If you couldn't figure out what is going wrong or would like to report an issue,
you can submit a simplified application that contains the problematic code using
the sandbox.

The `/sandbox` directory in the `loopback-next` monorepo can be used to utilize
the latest source code as symbolically-linked dependencies. See its
[README](https://github.com/strongloop/loopback-next/tree/master/sandbox) for
usage instructions.

To submit the sample, you can create a LoopBack application and copy over the

> sandbox files, then publish it to the Github.

Reviewer 2: I think this section is partially duplicating the content in https://loopback.io/doc/en/contrib/Reporting-issues.html#loopback-4x-bugs.

Have you considered keeping this section shorter and refer readers to https://loopback.io/doc/en/contrib/Reporting-issues.html#loopback-4x-bugs for more details?

Author: @bajtos https://loopback.io/doc/en/contrib/Reporting-issues.html#loopback-4x-bugs doesn't really cover using sandbox to reproduce the bug.

---
lang: en
title: 'Troubleshooting'
keywords: LoopBack 4.0, LoopBack 4, Node.js, TypeScript, Debug
sidebar: lb4_sidebar
permalink: /doc/en/lb4/Troubleshooting.html
---

In general, when you are developing an application, use the `npm` command to run
it. This enables you to see stack traces and console output immediately.

For example:

```
$ cd myapp
$ npm start
```

LoopBack 4 also provides multiple ways to help you verify or diagnose the app.

## Setting Debug Strings

> If you get an error but it doesn't provide enough information, you can turn on

Reviewer: Since we have a dedicated page for debug strings, perhaps we can simply the section by:

```
If you get an error but it doesn't provide enough information, you can turn on debug mode by starting your application using `DEBUG=<DEBUG_STRING> npm start`. [Setting debug strings](Setting-debug-strings.md) for details.
```

debug mode. That is, instead of starting the application using `npm start`, use
the following command:

```sh
DEBUG=loopback:* npm start
```

This command prints out all the debug information from LoopBack framework. To
only show a certain area's information, you can use more specific strings, see
page [Setting debug strings](Setting-debug-strings.md) for details.

## Running VS Code Tasks

> LoopBack team develops with editor Visual Studio Code in which we defined some

Reviewer: Suggested change:
```
If you're using Visual Studio Code, refer to [Debugging with VS Code page](Debugging-with-vscode.md) for debugging tips.
```

tasks to help people debug their projects. Each package in the monorepo contains
a `.vscode` folder for those tasks, see page
[Debugging with VS Code](Debugging-with-vscode.md) for details.

## Resolving Binding Error

With high extensibility, a LoopBack application usually contains tens of
bindings. A common error caused by missing bindings is "a binding key not
bound", for example:

```
500 Error: The key 'services.hasher' is not bound to any value in context application
```

To fix it, you should set the value for the binding key in the application.

For example, go to your application, typically it is in `src/application.ts`
file:

```ts
import {Hasher} from '/services/hasher';

export class TodoListApplication extends BootMixin(
  ServiceMixin(RepositoryMixin(RestApplication)),
) {
  constructor(options: ApplicationConfig = {}) {
    super(options);
    // Add your binding here
    this.bind(services.hasher).toClass(Hasher);
  }
}
```

See more details in
[Binding documentation page](https://loopback.io/doc/en/lb4/Binding.html).

## Updating Compiled Files

> The TypeScript application is compiled to JavaScript files and runs. Therefore

Reviewer: ```suggestion
The TypeScript application is compiled to JavaScript files and runs. Make sure you rebuild the project before re-running the application. By default, if you start the application using `npm start`, it rebuilds the project for you.
```

forgetting to rebuild the updated TS files could run into errors.

> For example, people deleted a Controller file but running `npm start` still

Reviewer: ```suggestion
However, in some cases, the transpiled JavaScript files do not get updated properly and it causes errors. For example, you deleted a Controller file but still get errors about the removed Controller during application startup, it implies the compiled JS files probably didn't get updated.
```

keeps complaining about it. This error implies the compiled JS files probably
didn't get updated.

To fix it, run `npm run clean` to delete all the compiled files. By doing so, it
forces the build to generate the compiled JS files next time when you start the
application using `npm start` or call `npm run build`.

> ## Leveraging Sandbox

Reviewer: _I have to confess that I'm not familiar with the use of Sandbox_ 
If it's about making your application using the latest code from loopback-next repo, perhaps we can modify the title better describe it?

Author: I think it's for user submitting issues, let me update accordingly.

---
lang: en
title: 'Troubleshooting'
keywords: LoopBack 4.0, LoopBack 4, Node.js, TypeScript, Debug
sidebar: lb4_sidebar
permalink: /doc/en/lb4/Troubleshooting.html
---

In general, when you are developing an application, use the `npm` command to run
it. This enables you to see stack traces and console output immediately.

For example:

```
$ cd myapp
$ npm start
```

LoopBack 4 also provides multiple ways to help you verify or diagnose the app.

## Setting Debug Strings

If you get an error but it doesn't provide enough information, you can turn on
debug mode by starting your application using `DEBUG=<DEBUG_STRING> npm start`.
[Setting debug strings](Setting-debug-strings.md) for details.

## Running VS Code Tasks

If you're using Visual Studio Code, refer to
[Debugging with VS Code page](Debugging-with-vscode.md) for debugging tips.

## Running Tests with Mocha

You can debug tests by running mocha commands as documented on page

> [Debugging with Mocha tests](Debugging-tests-with-mocha.md).

Author: @bajtos Thank you for review, I am following the instructions in story https://github.com/strongloop/loopback-next/issues/5701, and the list contains all these tips from different angles. That's why my understanding is this page wants to cover all the troubleshooting related topics.

> [Debugging with Mocha tests](Debugging-tests-with-mocha.md).

Reviewer 2: I see, it's unfortunate that #5701 is open to different interpretations.

The problem I see in your current version: as a LB user troubleshooting a problem in my application, when I open the debugging guide, I see instructions that do not apply to my situation - the launch tasks created by `lb4 app` are different, there is no `packages` folder in my project, etc.

In my view, most of the documentation at loopback.io is for LoopBack users (application and extension developers). 

Content for contributors (to loopback-next and other repositories) belongs to "contrib" section of our documentation, which is currently spread between https://loopback.io/doc/en/lb4/code-contrib-lb4.html and https://loopback.io/doc/en/contrib/.

I feel it's super important to make it clear which content is for whom, to avoid confusion and frustration.

I am proposing the following way forward:

- In the new pages you are adding to `Troubleshooting` section of the main docs, provide information how to debug LoopBack applications and extensions.
- Information on debugging loopback-next should go to https://github.com/strongloop/loopback-next/blob/master/docs/site/DEVELOPING.md. VSCode-specific information belongs to https://github.com/strongloop/loopback-next/blob/master/docs/site/Developing-with-vscode.md, see also https://github.com/strongloop/loopback-next/blob/master/docs/site/Developing-with-webstorm.md
- If you like, you can cross-link these different resources.

> [Debugging with Mocha tests](Debugging-tests-with-mocha.md).

Author: 👍 @bajtos I post my plan in https://github.com/strongloop/loopback-next/pull/6371#issuecomment-696811055 ptal, thanks

## Resolving Binding Error

With high extensibility, a LoopBack application usually contains tens of
bindings. A common error caused by missing bindings is "a binding key not
bound", for example:

```
500 Error: The key 'services.hasher' is not bound to any value in context application
```

> To fix it, you should set the value for the binding key in the application.

Author: @bajtos thank you for all the details. I captured most of them. 

"To fix it, you should set the value for the binding key in the application." --> Actually I run into this problem very often, so I decide to still keep it.
And add your tips :)

> To fix it, you should set the value for the binding key in the application.

Reviewer 2: Perfect 👍🏻

For example, go to your application, typically it is in `src/application.ts`
file:

```ts
import {Hasher} from '/services/hasher';

export class TodoListApplication extends BootMixin(
  ServiceMixin(RepositoryMixin(RestApplication)),
) {
  constructor(options: ApplicationConfig = {}) {
    super(options);
    // Add your binding here
    this.bind(services.hasher).toClass(Hasher);
  }
}
```

See more details in
[Binding documentation page](https://loopback.io/doc/en/lb4/Binding.html).

## Updating Compiled Files

> The TypeScript application is compiled to JavaScript files and runs. Make sure

Reviewer: ```suggestion
The TypeScript application is compiled to JavaScript files and runs. Forgetting to rebuild could run into errors. Make sure
```

you rebuild the project before re-running the application. By default, if you
start the application using `npm start`, it rebuilds the project for you.

> Forgetting to rebuild could run into errors. For example, you deleted a

Reviewer: To combine the two paragraphs: 

```suggestion
In some cases, the compiled JS files might not get updated properly even you build the project. For example, you deleted a Controller file but still get errors about the removed Controller during application startup. To fix it....
```

Controller file but still get errors about the removed Controller during
application startup, it implies the compiled JS files probably didn't get
updated.

To fix it, run `npm run clean` to delete all the compiled files. By doing so, it
forces the build to generate the compiled JS files next time when you start the
application using `npm start` or call `npm run build`.

> ## Submitting Problematic Code using Sandbox

Reviewer: Suggested change: 
```suggestion
## Submitting sample application for problem determination using Sandbox
```

> If you couldn't figure out what is going wrong or would like to report an issue,

Reviewer: Reading the sandbox README, the instruction is about how to get your LB app running against the latest LB4 code, so I guess it's not about submitting a sample app?

Author: No the readme is about utilize the latest source code. I think this paragraph says the purpose in the first sentence. Hmm any confusion point?

```
The `/sandbox` directory in the `loopback-next` monorepo can be used to utilize
the latest source code as symbolically-linked dependencies. See its README for
usage instructions.
```

Reviewer: that's what i meant. so it's not about submitting a sample app for problem determination, isn't it?

Author: I think it's for both. You first test and demo with the latest code, then submit the sample app. I added another paragraph talking about how to submit the app.
